tests/regression/test_projection.py

```diff
@@ -243,6 +243,57 @@ def test_projector_expression():
         Projector(expr, vo)
 
 
+@pytest.mark.parametrize('tensor', ['scalar', 'vector', 'tensor'])
+@pytest.mark.parametrize('same_fspace', [False, True])
+def test_projector_bcs(tensor, same_fspace):
+    mesh = UnitSquareMesh(2, 2)
+    x = SpatialCoordinate(mesh)
+    if tensor == 'scalar':
+        V = FunctionSpace(mesh, "CG", 1)
+        V_ho = FunctionSpace(mesh, "CG", 5)
+        bcs = [DirichletBC(V_ho, Constant(0.5), (1, 3)),
+               DirichletBC(V_ho, Constant(-0.5), (2, 4))]
+        fct = cos(x[0]*pi*2)*sin(x[1]*pi*2)
+
+    elif tensor == 'vector':
+        V = VectorFunctionSpace(mesh, "CG", 1)
+        V_ho = VectorFunctionSpace(mesh, "CG", 5)
+        bcs = [DirichletBC(V_ho, Constant((0.5, 0.5)), (1, 3)),
+               DirichletBC(V_ho, Constant((-0.5, -0.5)), (2, 4))]
+        fct = as_vector([cos(x[0]*pi*2)*sin(x[1]*pi*2),
+                         cos(x[0]*pi*2)*sin(x[1]*pi*2)])
+
+    elif tensor == 'tensor':
+        V = TensorFunctionSpace(mesh, "CG", 1)
+        V_ho = TensorFunctionSpace(mesh, "CG", 5)
+        bcs = [DirichletBC(V_ho, Constant(((0.5, 0.5),
+                                           (0.5, 0.5))), (1, 3)),
+               DirichletBC(V_ho, Constant(((-0.5, -0.5),
+                                           (-0.5, -0.5))), (2, 4))]
+        fct = as_tensor([[cos(x[0]*pi*2)*sin(x[1]*pi*2),
+                          cos(x[0]*pi*2)*sin(x[1]*pi*2)],
+                         [cos(x[0]*pi*2)*sin(x[1]*pi*2),
+                          cos(x[0]*pi*2)*sin(x[1]*pi*2)]])
+
+    if same_fspace:
+        v = Function(V_ho).project(fct)
+    else:
+        v = Function(V).project(fct)
+
+    ret = Function(V_ho)
+    projector = Projector(v, ret, bcs=bcs, solver_parameters={"ksp_type": "preonly",
+                                                              "pc_type": "lu"})
+    projector.project()
+
+    ref = Function(V_ho)
+    # The method "project" already handles bcs.
+    # So we use this as a reference to make sure the
+    # projector does the same thing (up to machine precision).
+    project(v, ref, bcs=bcs, solver_parameters={"ksp_type": "preonly",
+                                                "pc_type": "lu"})
```

Comment:
This is going to be a bad test soon, because I would like to replace the internals of `project` with the construction of a `Projector`...